six/modules/c++/six.sidd/tests/test_sidd_byte_provider.cpp

```diff
@@ -511,57 +550,70 @@ void Tester<DataTypeT>::testMultipleWritesBlocked()
     std::auto_ptr<const nitf::ImageBlocker> imageBlocker =
             siddByteProvider.getImageBlocker();
 
-    std::vector<DataTypeT> inData(mNumRowsPerBlock * mDims.col);
-
     const size_t numSegs(imageBlocker->getNumSegments());
-    for (size_t ii = 0; ii < numSegs; ++ii)
+    size_t totalNumBlocks(0);
+    for (size_t seg = 0; seg < numSegs; ++seg)
     {
-        const size_t seg = numSegs - ii - 1;
-        const size_t numBlocks = imageBlocker->getNumRowsOfBlocks(seg);
-        const size_t lastBlock = numBlocks - 1;
-        const size_t segStartRow = imageBlocker->getStartRow(seg);
-        const size_t numRowsPerBlock = imageBlocker->getNumRowsPerBlock()[seg];
-
-        std::vector<DataTypeT> blockData(
-                numRowsPerBlock * mNumColsPerBlock *
-                imageBlocker->getNumColsOfBlocks());
+        totalNumBlocks += imageBlocker->getNumRowsOfBlocks(seg);
+    }
 
-        for (size_t jj = 0; jj < numBlocks; ++jj)
+    size_t currentSegment = 0;
+    size_t numBlocksThisSeg = imageBlocker->getNumRowsOfBlocks(currentSegment);
+    size_t startBlockThisSeg = 0;
+    size_t startRow = 0;
+
+    // Want to write out the blocks in reverse order to make sure offsetting
+    // works. But I don't want to do the math backwards, so this will just go
+    // through and collect the data needed for each block write.
+    std::vector<WriteData> data;
+    for (size_t startBlock = 0;
+            startBlock < totalNumBlocks;
+            startBlock += blocksPerWrite)
+    {
+        WriteData writeData;
+        if (startBlock >= startBlockThisSeg + numBlocksThisSeg)
         {
-            const size_t block = lastBlock - jj;
+            ++currentSegment;
+            startBlockThisSeg += numBlocksThisSeg;
+            numBlocksThisSeg = imageBlocker->getNumRowsOfBlocks(currentSegment);
+        }
+        const size_t blocksThisWrite =
+                std::min<size_t>(blocksPerWrite, totalNumBlocks - startBlock);
 
-            const size_t startRow = segStartRow + block * numRowsPerBlock;
+        const size_t numRows = countNumRows(*imageBlocker, startBlock, blocksThisWrite,
+                currentSegment, startBlockThisSeg, numBlocksThisSeg);
 
-            size_t numRows(numRowsPerBlock);
-            if (block == lastBlock)
-            {
-                numRows -= imageBlocker->getNumPadRowsInFinalBlock(seg);
-            }
+        writeData.numBlocks = blocksThisWrite;
+        writeData.startRow = startRow;
+        writeData.numRows = numRows;
+        data.push_back(writeData);
+
+        startRow += numRows;
+    }
+
+    for (int ii = data.size() - 1; ii >= 0; --ii)
```

Comment:
Going to get a warning about signed/unsigned comparison here.  So instead, do something like...

```
for (size_t ctr = 0, ii = data.size() - 1; ctr < data.size(); ++ctr, --ii)
```